src/internal.c

```diff
@@ -6445,124 +6524,166 @@ int SendKexDhReply(WOLFSSH* ssh)
             } ecc;
 #endif
         } sk;
-    } sigKeyBlock;
+};
 
-    byte sig[512];
-    word32 sigSz = sizeof(sig);
-    word32 sigBlockSz;
+#define KEX_F_SIZE (256 + 1)
+#define KEX_SIG_SIZE (512)
 
-    word32 payloadSz;
+
+/* SendKexDhReply()
+ * It is also the funciton used for MSGID_KEXECDH_REPLY. The parameters
+ * are analogous between the two messages. Where MSGID_KEXDH_REPLY has
+ * server's public host key (K_S), f, and the signature of H;
+ * MSGID_KEXECDH_REPLY has K_S, the server'e ephemeral public key (Q_S),
+ * and the signature of H. This also applies to the GEX version of this.
+ * H is calculated the same for KEXDH and KEXECDH, and has some exceptions
+ * for GEXDH. */
+int SendKexDhReply(WOLFSSH* ssh)
+{
+    int ret = WS_SUCCESS;
+    byte *f_ptr = NULL, *sig_ptr = NULL;
+    byte *y_ptr = NULL, *r_ptr = NULL, *s_ptr = NULL;
     byte scratchLen[LENGTH_SZ];
+    word32 fSz = KEX_F_SIZE;
+    word32 sigSz = KEX_SIG_SIZE;
+    byte useEcc = 0;
+    byte fPad = 0;
+    byte kPad = 0;
+    word32 sigBlockSz;
+    word32 payloadSz;
     word32 scratch = 0;
     byte* output;
     word32 idx;
-    int ret = WS_SUCCESS;
     byte msgId = MSGID_KEXDH_REPLY;
     enum wc_HashType enmhashId;
+#ifndef WOLFSSH_NO_DH
+    const byte* primeGroup = NULL;
+    word32 primeGroupSz = 0;
+    const byte* generator = NULL;
+    word32 generatorSz = 0;
+    struct wolfSSH_sigKeyBlockFull *sigKeyBlock_ptr;
+#endif
+#ifdef WOLFSSH_SMALL_STACK
+    sigKeyBlock_ptr = WMALLOC(sizeof(struct wolfSSH_sigKeyBlockFull),
+            ssh->ctx->heap, DYNTYPE_PRIVKEY);
+    f_ptr = WMALLOC(KEX_F_SIZE, ssh->ctx->heap, DYNTYPE_BUFFER);
+    sig_ptr = WMALLOC(KEX_SIG_SIZE, ssh->ctx->heap, DYNTYPE_BUFFER);
+    if(sigKeyBlock_ptr == NULL || f_ptr == NULL || sig_ptr == NULL)
+        ret = WS_MEMORY_E;
```

Comment:
Checking again, I found indeed a return point in case of invalid argument that could cause a leak.
Should be fixed by removing the `return` at line 6586.

This function is also intended to have a single return point. All the pointers that may be allocated when used with SMALL_STACK are checked and freed at the bottom of the function (line: 7219)